File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.js
function solution(picks, minerals) {
    let answer = 2e9;
    let mineralsFatigue = [[1, 1, 1], [5, 1, 1], [25, 5, 1]];
    
    let mineralsInt = [];
    for(const m of minerals) {
        if(m === 'diamond')
            mineralsInt.push(0);
        else if(m === 'iron')
            mineralsInt.push(1);
        else if(m === 'stone')
            mineralsInt.push(2);
    }
    
    function digUp(cur, sum) {
        let save = 0;
        let picksFatigue = [ 0, 0, 0 ];
        
        if(cur >= minerals.length || (picks[0] === 0 && picks[1] === 0 && picks[2] === 0)) {
            answer = Math.min(answer, sum);
            return;
        }
        
        for(let i = cur; i < cur + 5; i++) {
            if(i >= minerals.length) {
                save = i;
                break;;
            }
            
            picksFatigue[0] += mineralsFatigue[0][mineralsInt[i]];
            picksFatigue[1] += mineralsFatigue[1][mineralsInt[i]];
            picksFatigue[2] += mineralsFatigue[2][mineralsInt[i]];
            
            save = i;
        }
        
        for(let i = 0; i < 3; i++) {
            if(picks[i] !== 0){
                picks[i]--;
                digUp(save + 1, sum + picksFatigue[i]);
                picks[i]++;
            }
        }
    }
    
    digUp(0, 0);
    
    return answer;
}

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int answer = 2e9;
int mineralFatigue[3][3] = {{1, 1, 1}, {5, 1, 1}, {25, 5, 1}};
vector<int> mineralInt;

void digUp(int cur, int sum, vector<int> &picks, vector<string> &minerals) {
    int digUpSum[3] = { 0, };
    int save = 0;
    
    if(minerals.size() <= cur || picks[0] + picks[1] + picks[2] == 0) {
        answer = min(answer, sum);
        return;
    }
    
    for(int i = cur; i < cur + 5; i++){
        if(i >= minerals.size()){
            save = i;
            break;
        }
        
        digUpSum[0] += mineralFatigue[0][mineralInt[i]];
        digUpSum[1] += mineralFatigue[1][mineralInt[i]];
        digUpSum[2] += mineralFatigue[2][mineralInt[i]];
        
        save = i;
    }
    
    for(int i = 0; i < picks.size(); i++) {
        if(picks[i] != 0){
            picks[i]--;
            digUp(save + 1, sum + digUpSum[i], picks, minerals);
            picks[i]++;
        }
    }
}

int solution(vector<int> picks, vector<string> minerals) {
        for(auto &m: minerals) {
            if (m == "diamond")
                mineralInt.push_back(0);
            else if (m == "iron")
                mineralInt.push_back(1);
            else if (m == "stone")
                mineralInt.push_back(2);
        }
    
    digUp(0, 0, picks, minerals);
    
    return answer;
}